File: src/lib/store.ts
import "server-only";

import { randomUUID } from "node:crypto";
import { mkdir, readFile, rename, writeFile } from "node:fs/promises";
import path from "node:path";

/**
 * Persistence for the three public forms.
 *
 * The site is read-mostly and the write volume for a regional agency is a
 * handful of enquiries a day, so a JSON file behind an in-process mutex is the
 * right amount of database: no driver, no migration, and the operator can read
 * the file with `cat`. Swapping in SQLite or Postgres later means
 * reimplementing this one module.
 *
 * Writes go through a temp file and an atomic rename, so a crash mid-write can
 * never leave a half-serialised record on disk.
 */

export type Kind = "enquiry" | "quote" | "application";
export type Status = "new" | "contacted" | "closed";

export type Submission = {
  id: string;
  kind: Kind;
  status: Status;
  createdAt: string;

  name: string;
  email?: string;
  phone: string;
  company?: string;
  subject?: string;
  message?: string;

  // Quote-specific.
  service?: string;
  siteType?: string;
  district?: string;
  headcount?: string;
  startWhen?: string;

  // Application-specific.
  vacancyId?: string;
  vacancyTitle?: string;
  age?: string;
  education?: string;
  experience?: string;

  // Request metadata, useful for spam triage.
  userAgent?: string;
  remoteIp?: string;
};

const DATA_FILE =
  process.env.NBSS_DATA ?? path.join(process.cwd(), "data", "submissions.json");

/**
 * Serialises every read-modify-write. Two form posts landing in the same tick
 * would otherwise both read the same array and one would overwrite the other.
 */
let queue: Promise<unknown> = Promise.resolve();

function withLock<T>(fn: () => Promise<T>): Promise<T> {
  const run = queue.then(fn, fn);
  // Keep the chain alive even if this operation rejects.
  queue = run.catch(() => undefined);
  return run;
}

async function readAll(): Promise<Submission[]> {
  try {
    const raw = await readFile(DATA_FILE, "utf8");
    return raw.trim() ? (JSON.parse(raw) as Submission[]) : [];
  } catch (err) {
    if ((err as NodeJS.ErrnoException).code === "ENOENT") return [];
    throw err;
  }
}

async function writeAll(subs: Submission[]): Promise<void> {
  await mkdir(path.dirname(DATA_FILE), { recursive: true });
  // A unique temp name keeps concurrent processes (e.g. a rebuild) from
  // clobbering each other's partial writes.
  const tmp = `${DATA_FILE}.${randomUUID()}.tmp`;
  await writeFile(tmp, JSON.stringify(subs, null, 2), "utf8");
  await rename(tmp, DATA_FILE);
}

/** Derives the next reference number from what is already on disk. */
function nextReference(subs: Submission[]): string {
  const highest = subs.reduce((max, s) => {
    const n = Number.parseInt(s.id.replace("NBSS-", ""), 10);
    return Number.isFinite(n) && n > max ? n : max;
  }, 0);
  return `NBSS-${String(highest + 1).padStart(5, "0")}`;
}

export type NewSubmission = Omit<Submission, "id" | "status" | "createdAt">;

/** Stores a submission and returns it with its reference number stamped on. */
export function addSubmission(input: NewSubmission): Promise<Submission> {
  return withLock(async () => {
    const subs = await readAll();
    const submission: Submission = {
      ...input,
      id: nextReference(subs),
      status: "new",
      createdAt: new Date().toISOString(),
    };
    subs.push(submission);
    await writeAll(subs);
    return submission;
  });
}

/** Newest first, optionally filtered by kind. */
export function listSubmissions(kind?: Kind): Promise<Submission[]> {
  return withLock(async () => {
    const subs = await readAll();
    return subs
      .filter((s) => !kind || s.kind === kind)
      .sort((a, b) => b.createdAt.localeCompare(a.createdAt));
  });
}

export function setStatus(id: string, status: Status): Promise<boolean> {
  return withLock(async () => {
    const subs = await readAll();
    const found = subs.find((s) => s.id === id);
    if (!found) return false;
    found.status = status;
    await writeAll(subs);
    return true;
  });
}

export type Counts = Record<string, number>;

export function countSubmissions(): Promise<Counts> {
  return withLock(async () => {
    const subs = await readAll();
    const counts: Counts = {
      total: subs.length,
      new: 0,
      enquiry: 0,
      quote: 0,
      application: 0,
    };
    for (const s of subs) {
      counts[s.kind] = (counts[s.kind] ?? 0) + 1;
      if (s.status === "new") counts.new += 1;
    }
    return counts;
  });
}

File: internal/store/store.go
// Package store persists the submissions that come off the public forms.
//
// The whole site is read-mostly, and the write volume for a regional security
// agency is a handful of enquiries a day. A JSON file guarded by a mutex is
// therefore the right amount of database: no driver, no migration, no cgo, and
// the operator can read the file with `cat`. Swapping this for SQLite or
// Postgres later means reimplementing this one interface.
package store

import (
	"encoding/json"
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"sync"
	"time"
)

// Kind distinguishes the three submission types sharing one file.
type Kind string

const (
	KindEnquiry     Kind = "enquiry"
	KindQuote       Kind = "quote"
	KindApplication Kind = "application"
)

// Status tracks where the sales or HR team has taken a submission.
type Status string

const (
	StatusNew      Status = "new"
	StatusContacted Status = "contacted"
	StatusClosed   Status = "closed"
)

// Submission is one record from any of the public forms. Fields that do not
// apply to a given Kind are simply left empty rather than modelled separately —
// the admin view is a single list and benefits from one shape.
type Submission struct {
	ID        string    `json:"id"`
	Kind      Kind      `json:"kind"`
	Status    Status    `json:"status"`
	CreatedAt time.Time `json:"created_at"`

	Name    string `json:"name"`
	Email   string `json:"email,omitempty"`
	Phone   string `json:"phone"`
	Company string `json:"company,omitempty"`
	Subject string `json:"subject,omitempty"`
	Message string `json:"message,omitempty"`

	// Quote-specific.
	Service   string `json:"service,omitempty"`
	SiteType  string `json:"site_type,omitempty"`
	District  string `json:"district,omitempty"`
	Headcount string `json:"headcount,omitempty"`
	StartWhen string `json:"start_when,omitempty"`

	// Application-specific.
	VacancyID    string `json:"vacancy_id,omitempty"`
	VacancyTitle string `json:"vacancy_title,omitempty"`
	Age          string `json:"age,omitempty"`
	Education    string `json:"education,omitempty"`
	Experience   string `json:"experience,omitempty"`

	// Request metadata, useful for spam triage.
	UserAgent string `json:"user_agent,omitempty"`
	RemoteIP  string `json:"remote_ip,omitempty"`
}

// Store is a concurrency-safe, file-backed collection of submissions.
type Store struct {
	mu   sync.RWMutex
	path string
	subs []Submission
	seq  int
}

// ErrNotFound is returned when an ID does not exist.
var ErrNotFound = errors.New("submission not found")

// Open loads the store at path, creating the parent directory and an empty file
// if they do not exist yet.
func Open(path string) (*Store, error) {
	if err := os.MkdirAll(filepath.Dir(path), 0o755); err != nil {
		return nil, fmt.Errorf("create data dir: %w", err)
	}
	s := &Store{path: path}

	b, err := os.ReadFile(path)
	switch {
	case errors.Is(err, os.ErrNotExist):
		return s, s.flush() // materialise an empty file so ops can see it exists
	case err != nil:
		return nil, fmt.Errorf("read %s: %w", path, err)
	}
	if len(b) > 0 {
		if err := json.Unmarshal(b, &s.subs); err != nil {
			return nil, fmt.Errorf("parse %s: %w", path, err)
		}
	}
	// Rebuild the sequence from what is on disk so IDs never collide after a
	// restart.
	for _, sub := range s.subs {
		var n int
		if _, err := fmt.Sscanf(sub.ID, "NBSS-%d", &n); err == nil && n > s.seq {
			s.seq = n
		}
	}
	return s, nil
}

// Add assigns an ID and timestamp, appends the submission and writes through to
// disk. The stored copy is returned so the caller can show the reference number.
func (s *Store) Add(sub Submission) (Submission, error) {
	s.mu.Lock()
	defer s.mu.Unlock()

	s.seq++
	sub.ID = fmt.Sprintf("NBSS-%05d", s.seq)
	sub.CreatedAt = time.Now()
	if sub.Status == "" {
		sub.Status = StatusNew
	}
	s.subs = append(s.subs, sub)

	if err := s.flush(); err != nil {
		// Roll back the in-memory append so memory and disk cannot diverge.
		s.subs = s.subs[:len(s.subs)-1]
		s.seq--
		return Submission{}, err
	}
	return sub, nil
}

// List returns submissions newest-first, optionally filtered by kind.
// An empty kind returns everything.
func (s *Store) List(kind Kind) []Submission {
	s.mu.RLock()
	defer s.mu.RUnlock()

	out := make([]Submission, 0, len(s.subs))
	for _, sub := range s.subs {
		if kind == "" || sub.Kind == kind {
			out = append(out, sub)
		}
	}
	sort.Slice(out, func(i, j int) bool { return out[i].CreatedAt.After(out[j].CreatedAt) })
	return out
}

// SetStatus moves a submission through the triage states.
func (s *Store) SetStatus(id string, status Status) error {
	s.mu.Lock()
	defer s.mu.Unlock()

	for i := range s.subs {
		if s.subs[i].ID != id {
			continue
		}
		prev := s.subs[i].Status
		s.subs[i].Status = status
		if err := s.flush(); err != nil {
			s.subs[i].Status = prev
			return err
		}
		return nil
	}
	return ErrNotFound
}

// Counts returns the number of submissions per kind, plus a "new" total.
func (s *Store) Counts() map[string]int {
	s.mu.RLock()
	defer s.mu.RUnlock()

	c := map[string]int{"total": len(s.subs)}
	for _, sub := range s.subs {
		c[string(sub.Kind)]++
		if sub.Status == StatusNew {
			c["new"]++
		}
	}
	return c
}

// flush writes the whole slice to disk atomically. Callers must hold the write
// lock. Rewriting the entire file is fine at this volume and removes any chance
// of a partially-appended record.
func (s *Store) flush() error {
	b, err := json.MarshalIndent(s.subs, "", "  ")
	if err != nil {
		return fmt.Errorf("encode submissions: %w", err)
	}
	tmp := s.path + ".tmp"
	if err := os.WriteFile(tmp, b, 0o644); err != nil {
		return fmt.Errorf("write temp file: %w", err)
	}
	if err := os.Rename(tmp, s.path); err != nil {
		return fmt.Errorf("replace %s: %w", s.path, err)
	}
	return nil
}
